File: android/src/game/object/Animal.java
package game.object;

import game.framework.Graphics;
import game.framework.impl.AndroidGame;
import game.util.Constrain;
import game.util.ImageResource;
import game.util.Utils;

public class Animal extends Object {
	
	public Animal(int type, int x, int y) {
		isGoUp = Utils.getRandom()%2 == 0 ? true : false;
		isGoLeft = Utils.getRandom()%2 == 0 ? true : false;
		setTypeItem(type);
		setXY(x, y);
		speed = Utils.getRandom(0, 4);
		speedUp = Utils.getRandom(0, 3);
		timeChangeAnimation = timeSeconds = System.currentTimeMillis();
		w = getImage().getWidth();
		h = getImage().getHeight();
	}

	public Animal(int type, int x, int y, int speed) {
		super(type, x, y, speed);
	}

	public Animal(int type, int x, int y, int w, int h) {
		super(type, x, y, w, h);
	}
	
	public Animal(int type, int x, int y, int w, int h, int speed) {
		super(type, x, y, w, h, speed);
	}
	
	public void setTypeItem(int type) {
		super.setTypeItem(typeItem);
		String strImage = Constrain.imgAnimal + type + ".png";
		setImage(ImageResource.getImage(strImage));
	}


	public void onPaint(Graphics g) {
		g.drawImage(getImage(), x, y, Graphics.LEFT);
	}
	
	public void onUpdate() {
		if(System.currentTimeMillis() - timeChangeAnimation > Utils.getRandom(1000, 2000) + 2000) {
			timeChangeAnimation = System.currentTimeMillis();
			int ran = Utils.getRandom(0, 100) % 4;
			if(ran > 0) {
				speed = Utils.getRandom(5, 10);
				speedUp = Utils.getRandom(5, 10) % 3;
				isGoLeft = Utils.getRandom()%2 == 0 ? true : false;
				isGoUp = Utils.getRandom()%2 == 0 ? true : false;
			}
		}
		if(isGoLeft) {
			if(x <= 0) {
				x = 0;
				isGoLeft = false;
			} else {
				x -= speed;
			}
		} else {
			if(x >= AndroidGame.widthScreen - getWidth()) {
				isGoLeft = true;
			} else {
				x += speed;
			}
		}
		
		if(isGoUp) {
			if(y <= 30) {
				y = 30;
				isGoUp = false;
			} else {
				y -= speedUp;
			}
		} else {
			if(y >= 200) {
				isGoUp = true;
			} else {
				y += speedUp;
			}
		}
		
	}

	public boolean onPorm() {
		long timeSeconds = System.currentTimeMillis();
		if(timeSeconds - this.timeSeconds >= Constrain.TIME_ANIMAL_PORN * 1000) {
			this.timeSeconds = timeSeconds;
			return true;
		}
		return false;
	}


	@Override
	public void onDestroy() {
		
	}

}

File: c++/Animal.cpp
#include "Animal.h"
#include "Constrain.h"
#include "Utils.h"

Animal::Animal(int type, int x, int y)
{
	isGoUp = rand()%2;
	isGoLeft = rand()%2;
	setType(type);
	setXY(x, y);
	speed = rand() % 4;
	speedUp = rand() % 3;
	timeChangeAnimation = timeSeconds = Utils::getTimeNow();
}

void Animal::onDestroy() {
	super::onDestroy();
}

void Animal::setType(int typeAnimal) {
	string strTmp = imgAnimal + to_string(typeAnimal) + ".png";
	char *str =(char *)strTmp.c_str();
	setTexture(str, 0, 0);
}

void Animal::onPaint() {
	super::onPaint();
}

void Animal::onUpdate() {
	if(Utils::getTimeNow() - timeChangeAnimation > rand()%3000 + 2000) {
		timeChangeAnimation = Utils::getTimeNow();
		int ran = rand() % 4;
		if(ran > 0) {
			speed = rand() % 4;
			speedUp = rand() % 3;
			isGoLeft = rand()%2;
			isGoUp = rand()%2;
		}
	}
	if(isGoLeft) {
		if(x <= 0) {
			x = 0;
			isGoLeft = FALSE;
		} else {
			x -= speed;
		}
	} else {
		if(x >= SCREEN_WIDTH - getWidth()) {
			isGoLeft = TRUE;
		} else {
			x += speed;
		}
	}
	
	if(isGoUp) {
		if(y <= 30) {
			y = 30;
			isGoUp = FALSE;
		} else {
			y -= speedUp;
		}
	} else {
		if(y >= 200) {
			isGoUp = TRUE;
		} else {
			y += speedUp;
		}
	}
	
}

bool Animal::onPorm() {
	long timeSeconds = Utils::getTimeNow();
	if(timeSeconds - this->timeSeconds >= TIME_ANIMAL_PORN * 1000) {
		this->timeSeconds = timeSeconds;
		return true;
	}
	return false;
}

void Animal::setTexture(char sFilename[], int width, int height) {
	super::setTexture(sFilename, width, height);
}

void Animal::setXY(int x, int y) {
	super::setXY(x, y);
}

void Animal::setX(int x) {
	super::setX(x);
}

void Animal::setY(int y) {
	super::setY(y);
}

int Animal::getX() {
	return super::getX();
}

int Animal::getY() {
	return super::getY();
}

int Animal::getWidth() {
	return super::getWidth();
}

int Animal::getHeight() {
	return super::getHeight();
}